### ISSUE TITLE
Categories mess

### ISSUE THREAD

**Issue by ksurent** [xsawyerx/pakket#43]
We need to decide on a few things when it comes package categories.

We currently allow invoking Pakket in the following way: `pakket build category/package`. Here, the user provides the category (meaning they need to know it beforehand). We also allow providing a text file with multiple packages in the same format: `category/package`. Again, the user needs to know categories beforehand.
The category value is used when computing paths to package configuration files.

At the same time we insist that configuration files themselves provide category. Which seems silly because we have to know it beforehand to be able to read configuration files in the first place. This is redundant information and a potential source of inconsistencies.

I suggest we remove the category field from package configs and cleanup related code. Wherever we need a category value, we use user supplied values.

Or, we could potentially remove the requirement of supplying package category beforehand but that would require making sure that package names are unique across all categories.
While both options would work, I think, the first one seems easier and the second one doesn't provide a lot of value.

**Comment by xsawyerx** [comment]
A full package is the name of it including the category of it. This is why you need to know it when you build it. That's perfectly fine and by design.

Having the category and name of package (and version) in its configuration is a matter of convenience and also on purpose. I don't know if it would serve much to remove it. (There is already [code](https://github.com/xsawyerx/pakket/blob/master/lib/Pakket/Builder.pm#L244) to assert the correctness of it.)

The idea is that by picking up a configuration you know what something is. If you remove it, a configuration depends on the storage layer. If it's a filesystem, you depend on directory *(category)* + directory *(package)* + filename *(version)*. If it's not a filesystem, but rather an ElasticSearch instance, you depend on mapping and further namespacing.

In short, I don't see how this is a problem.

**Comment by ksurent** [comment]
> The idea is that by picking up a configuration you know what something is. If you remove it, a configuration depends on the storage layer.

This is already the case is what I'm trying to say. Pakket itself, scaffolding scripts, everything is built with that assumption.
We can go along with that assumption and say it's a requirement now or we can change our code to remove it. Of course, we can also just do nothing.

> (There is already code to assert the correctness of it.)

We wouldn't need that code if we didn't have category in config files.
What does it mean when that assertion fails, anyway? Does it indicate a bug in Pakket?
To me it looks like there's a very good chance someone just typo-ed when manually editing config files.

> In short, I don't see how this is a problem.

We have two pieces of data. One of them is constantly used and always implicitly validated. The other one is not really used anywhere but is still a potential source of failed assertions.

**Comment by xsawyerx** [comment]
We've discussed this in person but I want to summarize this for prosperity.

By having each configuration self-contain the information about the package (name, category, version), we can then decouple the structure from the information. We can change the structure to use hashes (with an index), for example. We could then open the details (which might just as well be in some database) and find the information available. We can also support aliases this way.

Once we do this, we'll have to remove the code mentioned above because it forces the structure to reflect the data.
